ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStream.java

```diff
@@ -52,96 +60,196 @@ private TopicStream() {
+        keyFormatter = getKeyFormatter(records);
+      }
+
+      if (!valueFormatter.isPresent()) {
+        valueFormatter = getValueFormatter(records);
+      }
+
+      return StreamSupport.stream(records.spliterator(), false)
+          .map(this::delayedFormat)
           .collect(Collectors.toList());
     }
 
-    public Format getFormat() {
-      return formatter == null ? Format.UNDEFINED : formatter.getFormat();
+    @SuppressWarnings("OptionalGetWithoutIsPresent") // will not be empty if needed
+    private Supplier<String> delayedFormat(final ConsumerRecord<Bytes, Bytes> record) {
+      return () -> {
+        try {
+          final String rowTime = record.timestamp() == ConsumerRecord.NO_TIMESTAMP
+              ? "N/A"
+              : dateFormat.format(new Date(record.timestamp()));
+
+          final String rowKey = record.key() == null || record.key().get() == null
+              ? "<null>"
+              : keyFormatter.get().print(record.key());
+
+          final String value = record.value() == null || record.value().get() == null
+              ? "<null>"
+              : valueFormatter.get().print(record.value());
+
+          return "rowtime: " + rowTime
+              + ", " + "key: " + rowKey
+              + ", value: " + value;
+        } catch (IOException e) {
+          log.warn("Exception formatting record", e);
+          return "Failed to parse row";
+        }
+      };
+    }
+
+    public String getKeyFormat() {
+      return keyFormatter
+          .map(Formatter::getFormat)
+          .orElse(Format.UNDEFINED.toString());
+    }
+
+    public String getValueFormat() {
+      return valueFormatter
+          .map(Formatter::getFormat)
+          .orElse(Format.UNDEFINED.toString());
+    }
+
+    private Optional<Formatter> getKeyFormatter(
+        final Iterable<ConsumerRecord<Bytes, Bytes>> records
+    ) {
+      if (Iterables.isEmpty(records)) {
+        return Optional.empty();
+      }
+
+      final Stream<Bytes> valueStream = StreamSupport
+          .stream(records.spliterator(), false)
+          .map(ConsumerRecord::key);
+
+      return findFormatter(valueStream);
+    }
+
+    private Optional<Formatter> getValueFormatter(
+        final Iterable<ConsumerRecord<Bytes, Bytes>> records
+    ) {
+      if (Iterables.isEmpty(records)) {
+        return Optional.empty();
+      }
+
+      final Stream<Bytes> valueStream = StreamSupport
+          .stream(records.spliterator(), false)
+          .map(ConsumerRecord::value);
+
+      return findFormatter(valueStream);
+    }
+
+    private Optional<Formatter> findFormatter(final Stream<Bytes> dataStream) {
+      final List<Formatter> formatters = dataStream
+          .filter(Objects::nonNull)
+          .filter(d -> d.get() != null)
+          .map(this::findFormatter)
+          .collect(Collectors.toList());
+
+      final Set<String> formats = formatters.stream()
+          .map(Formatter::getFormat)
+          .collect(Collectors.toSet());
```

Comment:
I'm not sure I understand why we needed to turn this into a String?

Reply:
Because the format now has more info in it, e.g. `KAFKA (STRING)`.  The `(STRING)` is additional info.

However, again, this will change in the next PR.